socceranalyzer/common/analysis/goal_replay.py

```diff
@@ -0,0 +1,150 @@
+from socceranalyzer.common.enums.ssl import SSL
+from socceranalyzer.common.enums.vss import VSS
+from socceranalyzer.common.geometric.point import Point
+from socceranalyzer.common.chore.mediator import Mediator
+
+class GoalReplay:
+    """
+        Used to verify ball and players positions through N cycles before the occurrence of a goal.
+
+        GoalReplay(dataframe: pandas.DataFrame, category: enum, cycles: int)
+
+        Attributes
+        ----------
+            private:
+                dataframe: pandas.DataFrame
+                    match's log
+                category: SSL | SIM2D | VSS
+                    match's category
+                cycles: int
+                    number of cycles to be analyzed before the occurrence of each goal
+                ball_positions: [[Point]]
+                    a list containing a number of elements equivalent to the numbers of goals, with
+                    each element being itself a list containing the ball's positions for every cycle
+                    analyzed
+                left_team_positions: [[[Point]]]
+                    a list containing a number of elements equivalent to the numbers of goals, with
+                    each element being itself a list containing, for every cycle analyzed, another list
+                    corresponding to the left team players positions
+                right_team_positions: [[[Point]]]
+                    a list containing a number of elements equivalent to the numbers of goals, with
+                    each element being itself a list containing, for every cycle analyzed, another list
+                    corresponding to the right team players positions
+
+        Methods
+        -------
+            private:
+                find_goals() -> None
+                    Finds out at which cycles a goal happened and calls calculate().
+                calculate(starting_point: int) -> None
+                    Iterates backwards through cycles from the moment a goal happened and populates ball_positions, left_team_positions and
+                    right_team_positions using information from the dataframe.
+                    
+            public:
+                results() -> (ball_positions, left_team_positions, right_team_positions)
+                plot_ball(goal_number: int) -> None
+                    Plots ball positions obtained from calculate() onto a SIM2D field.
+    """
+
+    def __init__(self, dataframe: pandas.DataFrame, category: SSL | SIM2D | VSS, cycles: int) -> None:
+        self.__dataframe = dataframe
+        self.__category = category
+        self.__cycles = cycles
+        self.__ball_positions = []
+        self.__left_team_positions = []
+        self.__right_team_positions = []
+
+        self.__find_goals()
+    
+    def __find_goals(self) -> None:
+        """
+            Finds out at which cycles a goal happened and calls calculate().
+        """
+        for i in range(len(self.__dataframe)):
```

Comment:
Tentar ao máximo possível não utilizar **for** no dataframe, iterrows é uma alternativa.

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iterrows.html